Feature/gelato integration

superarius opened the pull request:
Hello Omen!

This pr is the last piece of Gelato's integration with Omen, allowing for Omen users to schedule an auto-withdraw of liquidity provided to a prediction market. Gelato handles transaction execution at the user specified auto-withdraw date, and the executor recoups the transaction fees by taking a portion of the liquidity withdrawn (the rate is calculated on chain using oracles for token and gas prices). As such we impose a minimum threshold on liquidity deposited before Gelato can be activated, since it doesn't make sense to use Gelato if the entire deposit amount being withdrawn automatically won't even cover the transaction fee (and we can only guess what the transaction fee will be at the future date of execution)

Also note that this PR works with an omen-subgraph that includes Gelato data. We have made a PR that is ready to be merged for that [here](https://github.com/protofire/omen-subgraph/pull/104)

Finally, there is a helpful constant for Omen devs `GELATO_ACTIVATED` that easily hides the Gelato features completely from the UI, which should be useful for e.g. xDai sidechain development (as Gelato services can only be used on main chain transactions)

Happy holidays!

hilmarx commented:
This PR implements #1028 @pimato

hilmarx commented:
Sneak Peak!

![2020-12-24 15 30 23](https://user-images.githubusercontent.com/19579480/103094185-f76b1f80-45fc-11eb-99a0-910ca98d4ad9.jpg)

superarius commented on app/src/common/constants.ts:
```diff
@@ -51,8 +51,8 @@ export const TRADING_FEE_OPTIONS = [
-export const GRAPH_MAINNET_WS = 'wss://api.thegraph.com/subgraphs/name/protofire/omen'
+export const GRAPH_MAINNET_HTTP = 'https://api.thegraph.com/subgraphs/name/hilmarx/gelato-omen-mainnet'
+export const GRAPH_MAINNET_WS = 'wss://api.thegraph.com/subgraphs/name/hilmarx/gelato-omen-mainnet'
 export const GRAPH_RINKEBY_HTTP = 'https://api.thegraph.com/subgraphs/name/hilmarx/omen-gelato'
```
Note: we changed the subgraph urls on mainnet and rinkeby to point to our deployed versions that include Gelato data, since our PR to omen-subgraph has not yet been merged. This is here so you can test everything on the UI properly now, once we merge the subgraph PR we'll change these urls back

KadenZipfel commented on app/src/services/gelato.ts:
```diff
@@ -0,0 +1,284 @@
+  'function cancelTask(tuple(uint256 id, address userProxy, tuple(address addr, address module) provider, uint256 index, tuple(tuple(address inst, bytes data)[] conditions, tuple(address addr, bytes data, uint8 operation, uint8 dataFlow, uint256 value, bool termsOkCheck)[] actions, uint256 selfProviderGasLimit, uint256 selfProviderGasPriceCeil)[] tasks, uint256 expiryDate, uint256 cycleId, uint25 ...
+]
+
+const actionWithdrawLiquidutyAbi = [
```
Can you fix the typo please? Should be 'actionWithdrawLiquidityAbi'

superarius replied:
done

KadenZipfel commented on app/src/services/gelato.ts:
```diff
@@ -0,0 +1,284 @@
+  }
+
+  /**
+   * Returns Uniswap price of collateral to DAI
```
Should we maybe use a more stable stablecoin, maybe USDC, to check the USD price? Or maybe there's a reasoning in favour of decentralization here?

hilmarx replied:
The assumption was that the difference in exchange rate that DAI and USDC will have will be negligible for the use case, but we can change it to USDC, I'm indifferent about this.

superarius replied:
Changed to USDC

KadenZipfel reviewed (changes requested):
Overall looks really good!

Mi-Lan reviewed (changes requested):
Very cool feature! Definitely needed, overall works great! Just found couple of small things that will be easy fix:
- As you can see in this screenshot seems that fields are not vertically aligned ![Screenshot 2020-12-29 at 11 48 34](https://user-images.githubusercontent.com/33226956/103278446-cbb8b280-49cb-11eb-9e36-52c9faed8337.png)
- Is it accounted for if market has duration that is shorter than 3 days?
- Also i don't know if it's scope of this issue but seems that "Your liquidity" in pool section stops working 
- Some parts of this text should be bolded like in desings
![Screenshot 2020-12-29 at 12 02 25](https://user-images.githubusercontent.com/33226956/103279324-cceadf00-49cd-11eb-9656-23335e42a5e5.png)
- This market is in test http://localhost:3000/#/0x915883aa30752c374527655192c2d9161d54e268 i created it with 12 dai via gelato and deposited 100 dai via gelato (this is just for future reference so i can test it)

superarius commented on app/src/services/cpk.ts:
```diff
@@ -574,6 +662,79 @@ class CPKService {
+    const depositAmount = Number(ethers.utils.formatUnits(amount, collateralToken.decimals))
+    const priorDepositAmount = Number(ethers.utils.formatUnits(priorAmount, collateralToken.decimals))
+    return {
+      belowMinimum: minDepositAmount > depositAmount + priorDepositAmount,
```
I get same issue but it's not actually from this line (here `belowMinimum` is true if `minDepositAmount` is strictly greater than total deposited i.e. we are NOT below the minimum if we deposit greater or equal to the `minDepositAmount` so its correct). The issue is coming from rounding issue (`minDepositAmount` is truncated on the UI but if I print all the decimals it says 'at least 100.00000001 DAI')

superarius replied:
^Sorry, above comment a bit confusing. What I meant was: when I tested putting exactly 100 DAI in the deposit field, it did not enable gelato as I would expect since the wording is "at least." But that wasn't because this inequality is backwards, it's because the minimum threshold being compared was not actually 100 DAI but in reality 100.00000001 DAI

superarius commented on app/src/components/market/common/gelato_scheduler/index.tsx:
```diff
@@ -0,0 +1,405 @@
+  if (!gelatoData.inputs) gelatoData.inputs = new Date(resolutionDateCopy.setDate(resolutionDateCopy.getDate() - 3))
+
+  const daysBeforeWithdraw = Math.round(
+    (Date.parse(resolution.toString()) - Date.parse(gelatoData.inputs.toString())) / 1000 / 60 / 60 / 24,
```
yes, done

superarius commented on app/src/components/market/common/gelato_scheduler/index.tsx:
```diff
@@ -0,0 +1,405 @@
+        : 'none'};
+    ${props => (props.filled ? 'fill: white;' : '')}
+  }
+  ${props => (props.filled ? 'background-color: #7986cb;' : '')}
```
thanks, done

superarius commented on app/src/components/common/icons/IconCheckmark.tsx:
```diff
@@ -0,0 +1,12 @@
+import React from 'react'
+
+export const IconCheckmark = () => (
```
Yes it is, we'll use that!

superarius commented on app/src/components/common/icons/IconCheckmarkFilled.tsx:
```diff
@@ -0,0 +1,14 @@
+import React from 'react'
+
+export const IconCheckmarkFilled = () => (
```
Roger

superarius commented:
> * As you can see in this screenshot seems that fields are not vertically aligned 

Fixed. Also FYI your screenshot has an icon that's a bit fishy (next to the word 'time' that icon should be hidden and is for me) maybe you are behind a few commits somehow?

> * Is it accounted for if market has duration that is shorter than 3 days?

We added that, thanks!

> * Also i don't know if it's scope of this issue but seems that "Your liquidity" in pool section stops working

Hmm... might be related to the first point but it should not do that. Is it that the amount of liquidity provided wont update or is it that a button won't activate on click?

> * Some parts of this text should be bolded like in desings

Done thanks.

Let me know if anything else is needed or if there are any issues with the changes. Thanks!

Mi-Lan commented:
> > * As you can see in this screenshot seems that fields are not vertically aligned
> 
> Fixed. Also FYI your screenshot has an icon that's a bit fishy (next to the word 'time' that icon should be hidden and is for me) maybe you are behind a few commits somehow?
> 
> > * Is it accounted for if market has duration that is shorter than 3 days?
> 
> We added that, thanks!
> 
> > * Also i don't know if it's scope of this issue but seems that "Your liquidity" in pool section stops working
> 
> Hmm... not sure either might also be related to the first point. Is it that the amount of liquidity displayed in "Your Liquidity" wont update or is it that a button won't activate on click?
> 
> > * Some parts of this text should be bolded like in desings
> 
> Done thanks.
> 
> Let me know if anything else is needed or if there are any issues with the changes. Thanks!

- Hmmm don't know yeah seems that icon is gone now!
- Regarding the buttons seems that they are still not quite yet center aligned just changing the wrapper element  to align-items: center seems to do the trick...
- Also the market i tested which i created with 12 dai and deposited 100 dai seemed to fail? Here is the market http://localhost:3001/#/0x915883aa30752c374527655192c2d9161d54e268
![Screenshot 2021-01-04 at 14 06 43](https://user-images.githubusercontent.com/33226956/103538212-189dfc80-4e96-11eb-8eef-57f950ef2641.png)
here is the transaction https://rinkeby.etherscan.io/tx/0x95a99d81c057e5a2c7a96e19063bdb90939b6d08d1f1102e568e109062edcd74#eventlog
What could be the issue there?

Mi-Lan reviewed (changes requested):
Also seems that this failed part  is not 100% according to design..
-here is the desgn
![Screenshot 2021-01-04 at 14 48 27](https://user-images.githubusercontent.com/33226956/103541737-23f42680-4e9c-11eb-9fa9-2303dc7c86a5.png)
-here it is in production
![Screenshot 2021-01-04 at 14 48 34](https://user-images.githubusercontent.com/33226956/103541770-2eaebb80-4e9c-11eb-8f2d-fc56d0b12762.png)

Mi-Lan reviewed (changes requested):
I found two things regarding gelato withdrawing and when market is resolved and auto-withdraw feature activates
- When i deposit liquidity with auto-withdraw and then withdraw before the period ends 
nothing changes in the ui below is ui after i withdraw
![Screenshot 2021-01-08 at 18 14 10](https://user-images.githubusercontent.com/33226956/104044578-c898a000-51dd-11eb-915b-13ca6a1f15e6.png)
after i refresh the page correct data is shown
![Screenshot 2021-01-08 at 18 14 33](https://user-images.githubusercontent.com/33226956/104044706-dcdc9d00-51dd-11eb-9be2-fbcdc9f6cf9c.png)
- When i stay in market view when liquidity is to be auto-withdrawn timer goes into negative and doesn't show correct data only on refresh does it change
![Screenshot 2021-01-08 at 12 51 17](https://user-images.githubusercontent.com/33226956/104044959-2927dd00-51de-11eb-8bcf-f4ed705d5f69.png)

superarius commented:
> I found two things regarding gelato withdrawing and when market is resolved and auto-withdraw feature activates
> 
> * When i deposit liquidity with auto-withdraw and then withdraw before the period ends
>   nothing changes in the ui below is ui after i withdraw
>   ![Screenshot 2021-01-08 at 18 14 10](https://user-images.githubusercontent.com/33226956/104044578-c898a000-51dd-11eb-915b-13ca6a1f15e6.png)
>   after i refresh the page correct data is shown
>   ![Screenshot 2021-01-08 at 18 14 33](https://user-images.githubusercontent.com/33226956/104044706-dcdc9d00-51dd-11eb-9be2-fbcdc9f6cf9c.png)

Hi we've pushed changes recently that should be fixing this (just tested and verified that after a cancel, it updated itself without refresh, it worked). It's possible that we were criss crossed a bit and you were still a commit or two behind, when you got that. The 'need to refresh' error should have been fixed by [this commit](https://github.com/protofire/omen-exchange/pull/1481/commits/a3d24817915e2fc8e0dee89917b9b28b18d49885). There is a chance that A) you were a commit or two behind (no worries since we pushed very recently) or B) subgraph was just being very slow to update?

Mi-Lan commented:
> > I found two things regarding gelato withdrawing and when market is resolved and auto-withdraw feature activates
> > 
> > * When i deposit liquidity with auto-withdraw and then withdraw before the period ends
> >   nothing changes in the ui below is ui after i withdraw
> >   ![Screenshot 2021-01-08 at 18 14 10](https://user-images.githubusercontent.com/33226956/104044578-c898a000-51dd-11eb-915b-13ca6a1f15e6.png)
> >   after i refresh the page correct data is shown
> >   ![Screenshot 2021-01-08 at 18 14 33](https://user-images.githubusercontent.com/33226956/104044706-dcdc9d00-51dd-11eb-9be2-fbcdc9f6cf9c.png)
> 
> Hi we've pushed changes recently that should be fixing this (just tested and verified that after a cancel, it updated itself without refresh, it worked). It's possible that we were criss crossed a bit and you were still a commit or two behind, when you got that. The 'need to refresh' error should have been fixed by [this commit](https://github.com/protofire/omen-exchange/pull/1481/commits/a3d24817915e2fc8e0dee89917b9b28b18d49885). There is a chance that A) you were a commit or two behind (no worries since we pushed very recently) or B) subgraph was just being very slow to update?

Yeah it is that i was testing it this afternoon!

Mi-Lan commented:
> > > I found two things regarding gelato withdrawing and when market is resolved and auto-withdraw feature activates
> > > 
> > > * When i deposit liquidity with auto-withdraw and then withdraw before the period ends
> > >   nothing changes in the ui below is ui after i withdraw
> > >   ![Screenshot 2021-01-08 at 18 14 10](https://user-images.githubusercontent.com/33226956/104044578-c898a000-51dd-11eb-915b-13ca6a1f15e6.png)
> > >   after i refresh the page correct data is shown
> > >   ![Screenshot 2021-01-08 at 18 14 33](https://user-images.githubusercontent.com/33226956/104044706-dcdc9d00-51dd-11eb-9be2-fbcdc9f6cf9c.png)
> > 
> > 
> > Hi we've pushed changes recently that should be fixing this (just tested and verified that after a cancel, it updated itself without refresh, it worked). It's possible that we were criss crossed a bit and you were still a commit or two behind, when you got that. The 'need to refresh' error should have been fixed by [this commit](https://github.com/protofire/omen-exchange/pull/1481/commits/a3d24817915e2fc8e0dee89917b9b28b18d49885). There is a chance that A) you were a commit or two behind (no worries since we pushed very recently) or B) subgraph was just being very slow to update?
> 
> Yeah it is that i was testing it this afternoon!



> > > I found two things regarding gelato withdrawing and when market is resolved and auto-withdraw feature activates
> > > 
> > > * When i deposit liquidity with auto-withdraw and then withdraw before the period ends
> > >   nothing changes in the ui below is ui after i withdraw
> > >   ![Screenshot 2021-01-08 at 18 14 10](https://user-images.githubusercontent.com/33226956/104044578-c898a000-51dd-11eb-915b-13ca6a1f15e6.png)
> > >   after i refresh the page correct data is shown
> > >   ![Screenshot 2021-01-08 at 18 14 33](https://user-images.githubusercontent.com/33226956/104044706-dcdc9d00-51dd-11eb-9be2-fbcdc9f6cf9c.png)
> > 
> > 
> > Hi we've pushed changes recently that should be fixing this (just tested and verified that after a cancel, it updated itself without refresh, it worked). It's possible that we were criss crossed a bit and you were still a commit or two behind, when you got that. The 'need to refresh' error should have been fixed by [this commit](https://github.com/protofire/omen-exchange/pull/1481/commits/a3d24817915e2fc8e0dee89917b9b28b18d49885). There is a chance that A) you were a commit or two behind (no worries since we pushed very recently) or B) subgraph was just being very slow to update?
> 
> Yeah it is that i was testing it this afternoon!
I seem to be getting this error when it comes to time to autho-withdraw....
![Screenshot 2021-01-08 at 19 22 34](https://user-images.githubusercontent.com/33226956/104050527-0c43d780-51e7-11eb-8843-813fcea6fb5f.png)

hilmarx commented:
> Yeah it is that i was testing it this afternoon!
> I seem to be getting this error when it comes to time to autho-withdraw....

That seems to be general issue with Apollo, I have encountered that one before on other dApps, dont think it has smth to do with Gelato. 

https://github.com/apollographql/react-apollo/issues/2813
or
https://github.com/apollographql/react-apollo/issues/763

Mi-Lan commented:
> > Yeah it is that i was testing it this afternoon!
> > I seem to be getting this error when it comes to time to autho-withdraw....
> 
> That seems to be general issue with Apollo, I have encountered that one before on other dApps, dont think it has smth to do with Gelato.
> 
> [apollographql/react-apollo#2813](https://github.com/apollographql/react-apollo/issues/2813)
> or
> [apollographql/react-apollo#763](https://github.com/apollographql/react-apollo/issues/763)

Yeah it seems to be the case!

superarius commented:
This repository is now up to date after rebase and commit to fix small new issues now that scalar markets have been integrated.

Note that currently, Gelato auto-withdrawls will only be available for categorical markets. In my opinion, integrating Gelato for scalar markets is outside the scope of this pr and can be handled separately but let me know what y'all think.

Lastly, we're doing the last minor styling corrections [in this pr](https://github.com/gelatodigital/omen-exchange/pull/6) so these changes are easier to find and discuss.

superarius commented:
Pushed another rebase, and updated some ethereum addresses and config urls for production. Just a note that if you want to test gelato on Mainnet, we have reinstated the graphql urls to point to the protofire subgraph, but it seems only rinkeby subgraph has been deployed and synced to include gelato. Either sync the new mainnet subgraph (that includes gelato) before testing, or just reset these two variables in the `constants.ts` :

export const GRAPH_MAINNET_HTTP = 'https://api.thegraph.com/subgraphs/name/hilmarx/gelato-omen-mainnet'
export const GRAPH_MAINNET_WS = 'wss://api.thegraph.com/subgraphs/name/hilmarx/gelato-omen-mainnet'

superarius commented:
closing in favor of #1691
